pkg/util/parse_numeric.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2021 The KEDA Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+)
+
+var hintedRegexp *regexp.Regexp
+var floatRegexp *regexp.Regexp
+var intRegexp *regexp.Regexp
+
+func init() {
+	hintedRegexp = regexp.MustCompile(`^(?P<Type>[a-z])\((?P<Value>[\-]?(?:(?:0|[1-9]\d*)(?:\.\d*)?|\.\d+))\)$`)
+	floatRegexp = regexp.MustCompile(`^-?\d+\.\d+$`)
+	intRegexp = regexp.MustCompile(`^-?\d+$`)
+}
+
+type typeHintError struct {
+	value string
+}
+
+func (e typeHintError) Error() string {
+	return fmt.Sprintf("ParseNumeric: Type hint \"%s\" does not match any known hints", e.value)
+}
+
+type numericParseError struct {
+	value string
+}
+
+func (e numericParseError) Error() string {
+	return fmt.Sprintf("ParseNumeric: Provided string \"%s\" is neither an int nor a float", e.value)
+}
+
+func ParseNumeric(s string, bitSize int) (interface{}, error) {
+	/* This function is a wrapper around ParseInt and ParseFloat
+	mostly, but also allows for providing in-string type hinting
+	to ensure that numeric values that would be cast to numbers
+	by YAML parsers instead forcefully remain as strings until
+	they can get parsed at runtime.
+
+	For this purpose, we look for a hint which is defined as
+	a single letter, followed by a parentheses wrapped numeric
+	value.  d or f for decimal/float, i or n for integer.
+
+	If we don't get a match, check if s looks like a float and
+	parse it as one if so, otherwise parse as an int.
+	*/
```

Comment:
These comments would be great as standard Go documentation:

```suggestion
// ParseNumeric is a wrapper around ParseInt and ParseFloat
// mostly, but also allows for providing in-string type hinting
// to ensure that numeric values that would be cast to numbers
// by YAML parsers instead forcefully remain as strings until
// they can get parsed at runtime.
//
// For this purpose, we look for a hint which is defined as
// a single letter, followed by a parentheses wrapped numeric
// value.  d or f for decimal/float, i or n for integer.
//
// If we don't get a match, check if s looks like a float and
// parse it as one if so, otherwise parse as an int.
func ParseNumeric(s string, bitSize int) (interface{}, error) {
```